core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java

```diff
@@ -11413,6 +11413,14 @@ private void checkCustomColumnResolving(String table) {
     expr("json_value('{\"foo\":100}', 'lax $.foo' returning boolean"
         + " default 100 on empty)")
         .columnType("BOOLEAN");
+
+    expr("json_value('{\"foo\":[100, null, 200]}', 'lax $.foo'"
+        + "returning integer array)")
+        .columnType("INTEGER ARRAY");
+
+    expr("json_value('{\"foo\":[[100, null, 200]]}', 'lax $.foo'"
```

Comment:
I am not familiar with this function. Can it be used when the array element types are not the same? JSON certainly allows that.

Reply:
I am unsure exactly of the standard; I can only speak for the behavior in Apache Druid where we have a dedicated json type rather than operating on strings and we coerce array elements to a common 'least restrictive type'.

Reply:
This has been a desired feature in Calcite for quite some time: https://issues.apache.org/jira/browse/CALCITE-4918
Does anyone plan to contribute it (assuming you are using Calcite for this purpose)?
I was thinking to do it myself, but if you already have the code and the tests, it would be great.

Reply:
Yeah, I agree it would be nice to have, but don't really have any pre-existing code for the calcite side of things for this. Druid could probably make use of it if such a thing exists as we store this type of varying json data in their original types in our segments, but right now at query time we best effort convert them to a common type for query processing.